### PR TITLE
fix: add support for strategy field to indicate how effective value was determined

### PR DESCRIPTION
**Pull request by leninmehedy**
## Description

This pull request introduces a new `EffectiveValue` type to the codebase, providing a structured way to represent both the value and the strategy used to compute the "effective" value in the `RuntimeValue` container. The changes refactor the `EffectiveFunc` and related methods to use this new type, update the caching and concurrency logic, and add comprehensive tests for the new behavior.

Key changes include:

**Core Type and API Changes:**

* Introduced the `EffectiveValue` type and `EffectiveStrategy` enum in a new file (`type_effective_value.go`), encapsulating both the value and the strategy (default, user input, or custom logic) used to compute it. Includes JSON marshaling and string conversion.
* Refactored the `EffectiveFunc` signature and related fields/methods in `runtime_value.go` to use `*EffectiveValue[T]` instead of `Value[T]`, updating all usages and cache logic accordingly. [[1]](diffhunk://#diff-735149d89d66dd98a8dcfcbcca658d494eb530bfddd0ba42b6543a7f6cab2f96L149-R149) [[2]](diffhunk://#diff-735149d89d66dd98a8dcfcbcca658d494eb530bfddd0ba42b6543a7f6cab2f96L173-R173) [[3]](diffhunk://#diff-735149d89d66dd98a8dcfcbcca658d494eb530bfddd0ba42b6543a7f6cab2f96L204-R204) [[4]](diffhunk://#diff-735149d89d66dd98a8dcfcbcca658d494eb530bfddd0ba42b6543a7f6cab2f96L254-R254) [[5]](diffhunk://#diff-735149d89d66dd98a8dcfcbcca658d494eb530bfddd0ba42b6543a7f6cab2f96L265-R265) [[6]](diffhunk://#diff-735149d89d66dd98a8dcfcbcca658d494eb530bfddd0ba42b6543a7f6cab2f96L322-R323) [[7]](diffhunk://#diff-735149d89d66dd98a8dcfcbcca658d494eb530bfddd0ba42b6543a7f6cab2f96L443-R449)

**Behavioral and Logic Updates:**

* Updated the default `effectiveFunc` to wrap values in `EffectiveValue` and assign the correct strategy, ensuring consistent strategy tracking and simplifying downstream logic.
* Adjusted all test cases in `runtime_value_test.go` to use the new `EffectiveValue` API, including checks for strategy and value access. [[1]](diffhunk://#diff-d6049c83d49fac755be39b20dd741b15eb3d1cfd9f34531c06de830378245d9dL146-R147) [[2]](diffhunk://#diff-d6049c83d49fac755be39b20dd741b15eb3d1cfd9f34531c06de830378245d9dL176-R177) [[3]](diffhunk://#diff-d6049c83d49fac755be39b20dd741b15eb3d1cfd9f34531c06de830378245d9dL189-R196) [[4]](diffhunk://#diff-d6049c83d49fac755be39b20dd741b15eb3d1cfd9f34531c06de830378245d9dL215-R225) [[5]](diffhunk://#diff-d6049c83d49fac755be39b20dd741b15eb3d1cfd9f34531c06de830378245d9dL239-R240) [[6]](diffhunk://#diff-d6049c83d49fac755be39b20dd741b15eb3d1cfd9f34531c06de830378245d9dL260-R260) [[7]](diffhunk://#diff-d6049c83d49fac755be39b20dd741b15eb3d1cfd9f34531c06de830378245d9dL269-R269) [[8]](diffhunk://#diff-d6049c83d49fac755be39b20dd741b15eb3d1cfd9f34531c06de830378245d9dL283-R283) [[9]](diffhunk://#diff-d6049c83d49fac755be39b20dd741b15eb3d1cfd9f34531c06de830378245d9dL309-R310) [[10]](diffhunk://#diff-d6049c83d49fac755be39b20dd741b15eb3d1cfd9f34531c06de830378245d9dL328-R328) [[11]](diffhunk://#diff-d6049c83d49fac755be39b20dd741b15eb3d1cfd9f34531c06de830378245d9dL349-R376) [[12]](diffhunk://#diff-d6049c83d49fac755be39b20dd741b15eb3d1cfd9f34531c06de830378245d9dL387-R411) [[13]](diffhunk://#diff-d6049c83d49fac755be39b20dd741b15eb3d1cfd9f34531c06de830378245d9dL422-R437) [[14]](diffhunk://#diff-d6049c83d49fac755be39b20dd741b15eb3d1cfd9f34531c06de830378245d9dL486-R489) [[15]](diffhunk://#diff-d6049c83d49fac755be39b20dd741b15eb3d1cfd9f34531c06de830378245d9dL498-R504)

**Testing:**

* Added a new test file, `type_effective_value_test.go`, to verify the behavior of `EffectiveValue` and `EffectiveStrategy`, including JSON marshaling, cloning, and helper functions.

These changes improve the clarity, extensibility, and correctness of how effective values and their provenance are managed throughout the system.

### Related Issues

* Closes #
* Related to #

### Pull request (PR) checklist

* \[ ] This PR added tests (unit, integration, and/or end-to-end)
* \[ ] This PR updated documentation
* \[ ] This PR added no TODOs or commented out code
* \[ ] This PR has no breaking changes
* \[ ] Any technical debt has been documented as a separate issue and linked to this PR
* \[ ] Any `go.mod` and/or `go.sum` changes have been explained to and approved by a repository manager
* \[ ] All related issues have been linked to this PR
* \[ ] All changes in this PR are included in the description
* \[ ] When this PR merges the commits will be squashed and the title will be used as the commit message, the 'commit
  message guidelines' below have been followed

### Testing

* \[ ] This PR added unit tests
* \[ ] This PR added integration/end-to-end tests
* \[ ] These changes required manual testing that is documented below
* \[ ] Anything not tested is documented

The following manual testing was done:

* TBD

The following was not tested:

* TBD

<details>
<summary>
Commit message guidelines
</summary>
We use 'Conventional Commits' to ensure that our commit messages are easy to read, follow a consistent format, and for automated release note generation. Please follow the guidelines below when writing your commit messages:

1. BREAKING CHANGE: a commit that has a footer BREAKING CHANGE:, or appends a ! after the type/scope, introduces a
   breaking API change (correlating with MAJOR in Semantic Versioning). A BREAKING CHANGE can be part of commits of any
   type. NOTE: currently breaking changes will only bump the MAJOR version.
2. The title is prefixed with one of the following:

| Prefix    | Description                                         | Semantic Version Update | Captured in Release Notes |
|-----------|-----------------------------------------------------|-------------------------|---------------------------|
| feat:     | a new feature                                       | MINOR                   | Yes                       |
| fix:      | a bug fix                                           | PATCH                   | Yes                       |
| perf:     | performance                                         | PATCH                   | Yes                       |
| refactor: | code change that isn't feature or fix               | none                    | No                        |
| test:     | adding missing tests                                | none                    | No                        |
| docs:     | changes to documentation                            | none                    | Yes                       |
| build:    | changes to build process                            | none                    | No                        |
| ci:       | changes to CI configuration                         | none                    | No                        |
| style:    | formatting, missing semi-colons, etc                | none                    | No                        |
| chore:    | updating grunt tasks etc; no production code change | none                    | No                        |

</details>

### CODE REVIEW

File: type_effective_value_test.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	assert.Equal(t, StrategyCustom, cev.Strategy())
+
+	// Clone should succeed and preserve strategy
+	clone, err := ev.Clone()
+	require.NoError(t, err)
+	require.NotNil(t, clone)
+	assert.Equal(t, ev.Strategy(), clone.Strategy())
+	// underlying value must be present on clone
+	assert.NotNil(t, clone.Get())
+}
```
[the remark this answers is not in the document]
Comment (leninmehedy) [reply type_effective_value_test.go]: nil isn't accepted anymore
